LoginClient/LoginClientForm.cs

```diff
@@ -0,0 +1,232 @@
+/// into a central server to test user and password combinations 
+/// (you must have at least one administrator password setup)
+/// You must create a Server Client program it must use IPC to communicate.
+/// Your program must have a login that uses standard hashing techniques.
+/// 
+/// </summary>
+namespace LoginClient
+{
+    public partial class LoginClientForm : Form
+    {
+        private PipeClient pipeClient;
+
+        public LoginClientForm()
+        {
+            InitializeComponent();
+        }
+
+        /// <summary>
+        /// This method sets up the program when it first loads.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void LoginClientForm_Load(object sender, EventArgs e)
+        {
+            CreateNewPipeClient();
+        }
+
+        /// <summary>
+        /// Create a new pipeClient and set all the handlers for the delegates.
+        /// </summary>
+        private void CreateNewPipeClient()
+        {
+            if (pipeClient != null)
+            {
+                pipeClient.Disconnect();
+                pipeClient.MessageRecieved -= pipeClient_MessageRecieved;
+                pipeClient.ServerDisconnected -= pipeClient_ServerDisconnected;
+                pipeClient.UpdateTheForm -= pipeClient_UpdateTheForm;
+                pipeClient = null;
+            }
+
+            pipeClient = new PipeClient();
+            pipeClient.MessageRecieved += pipeClient_MessageRecieved;
+            pipeClient.ServerDisconnected += pipeClient_ServerDisconnected;
+            pipeClient.UpdateTheForm += pipeClient_UpdateTheForm;
+
+            UpdateForm();
+        }
+
+        /// <summary>
+        /// Methods to allow the pipeClient to use the methods on the form.
+        /// </summary>
+        #region HandlerMethods
+
+        private void pipeClient_UpdateTheForm()
+        {
+            Invoke(new PipeClient.UpdateTheFormHandler(UpdateForm));
+        }
+
+        private void pipeClient_MessageRecieved(byte[] message)
+        {
+            Invoke(new PipeClient.MessageRecievedHandler(DisplayRecievedMessage),
+                new object[] { message });
+        }
+
+        private void pipeClient_ServerDisconnected()
+        {
+            Invoke(new PipeClient.ServerDisconnectedHandler(UpdateForm));
+        }
+        #endregion
+
+        /// <summary>
+        /// Add the message onto a newline of the message log.
+        /// </summary>
+        /// <param name="message"></param>
+        private void DisplayRecievedMessage(byte[] message)
+        {
+            Console.WriteLine(Thread.CurrentThread.ManagedThreadId + "in DisplayRecievedMessage");
+            ASCIIEncoding encoder = new ASCIIEncoding();
+            string str = encoder.GetString(message, 0, message.Length);
+
+            if (str == "close")
+            {
+                //MessageLogTB.Text += str + "\r\n";
+                pipeClient.Disconnect();
```

Comment:
Thought.
disconnect here, and in createnew.
This can be a sign, the function createnew, is confusing. since it also disconnects. 

general solution might be to create a reconnect function for when you need to disconnect and reconnect.
a disconnect function.
a connect function.

reconnect would then just be:
disconnect()
connect()